### ISSUE TITLE
Add coverage to tests

### ISSUE THREAD

**Issue by matheusgomes28** [matheusgomes28/base64pp#55]
Add coverage reporting to unit tests. Difficult task!

TODOS 📑 

- [x] Compile with `--coverage` in `gcc`. 📔 
- [x] Integrate `gcov` and `lcov` to read the profiling data. 👓 
- [x] Deploy coverage HTML to artifacts. 📫

**Comment by matheusgomes28** [comment]
Alternatively to `gcov` and `lcov`, you can integrate with `codecov`.

The plan is:

- [ ] Build debug with coverage flags, `--coverage` suffices.
- [ ] Calling `lcov` after the tests run: `lcov --capture --directory . --output-file coverage.info` from project dir suffices.
- [ ] Remove the stl and conan files with `lcov --remove coverage.info '/usr/*' --output-file coverage.info`

Uploading the coverage info:
- [ ] Create pipeline secret with Codecov PAT
- [ ] Upload the coverage info with `bash <(curl -s https://codecov.io/bash) -f coverage.info || echo "Codecov did not collect coverage reports"`

**Comment by matheusgomes28** [comment]
Perhaps the following would be necessary if there are issues with lcov finding the source files:

```
lcov -d BUILD/app/packages/ --capture --no-external --output-file app.info
     --rc geninfo_adjust_src_path="${PROJECT_ROOT}/build/unix-deb
     => ${PROJECT_ROOT}"
```

**Comment by matheusgomes28** [comment]
Turns out the docker container does not have lcov. Make sure we add this in, research a good lcov version to add.

**Comment by matheusgomes28** [comment]
Long but closed with #61
